### PR TITLE
osd: add smartctl error output to device map

### PR DESCRIPTION
**Pull request by yaarith**
In case smartctl returns an exit code other than 0, we include its
output on the device map, so that the mgr can decide how to handle it.

Signed-off-by: Yaarit Hatuka yaarithatuka@gmail.com

### CODE REVIEW

File: src/osd/OSD.cc [hunk 1]
```diff
@@ -6264,23 +6265,42 @@ int OSD::probe_smart_device(const char *device, int timeout, std::string *result
 
   int ret = smartctl.spawn();
   if (ret != 0) {
-    derr << "failed run smartctl: " << smartctl.err() << dendl;
-    return ret;
+    std::string err_msg = "failed to run smartctl: " + smartctl.err();
+    derr << err_msg << dendl;
+    *result = err_msg;
+    return -1;
```
Comment (liewegas) [inline src/osd/OSD.cc]: I think that here we should return an error.  Pretty sure the spawn means we start the subprocess, and it's the join() where we get the final return value.

File: src/osd/OSD.cc [hunk 2]
```diff
@@ -6264,23 +6265,42 @@ int OSD::probe_smart_device(const char *device, int timeout, std::string *result
...
   bufferlist output;
   ret = output.read_fd(smartctl.get_stdout(), 100*1024);
   if (ret < 0) {
-    derr << "failed read from smartctl: " << cpp_strerror(-ret) << dendl;
+    std::string err_msg = "failed to read from smartctl: " + cpp_strerror(-ret);  
+    derr << err_msg << dendl;
+    *result = err_msg;
   } else {
-    ret = 0;
     *result = output.to_str();
```
Comment (liewegas) [inline src/osd/OSD.cc]: I think this part can remain the same too... if we fail to read the output that is not good.

File: src/osd/OSD.cc [hunk 3]
```diff
@@ -6231,9 +6231,10 @@ void OSD::probe_smart(const string& only_devid, ostream& ss)
     }
 
     std::string result;
-    if (probe_smart_device(("/dev/" + dev).c_str(), smart_timeout, &result)) {
-      dout(10) << "probe_smart_device failed for /dev/" << dev << dendl;
-      //continue;
+    // Since we want to assign any smartctl result to json_map, regardless 
+    // of success or failure, we only check the result's length ,and ignore the return val.
+    probe_smart_device(("/dev/" + dev).c_str(), smart_timeout, &result);
+    if (!result.length()) { // no error message was assigned via probe_smart_device()
```
Comment (liewegas) [inline src/osd/OSD.cc]: I think this can stay the same...

File: src/osd/OSD.cc [hunk 4]
```diff
@@ -6264,23 +6265,42 @@ int OSD::probe_smart_device(const char *device, int timeout, std::string *result
...
+    or device is in a low-power mode.
+  (ret == 3) is not expected (due to bitmask).
+  The rest of smartctl exit status suggest there might be something wrong with the device,
+  but the output might still be useful. When (0 < ret < 4)
+  there is no actual output about the device (such as attr table, etc.), 
+  and we need to check what went wrong.
+  */
+  if (ret != 0) {
     derr << smartctl.err() << dendl;
-    return -EINVAL;
```
Comment (liewegas) [inline src/osd/OSD.cc]: I think this is the only piece that needs to change. Maybe add in here that if result is empty (smartctl returned error *and* failed to produce some output) then we set result to { "error": "smartctl returned error", "exit_code": r } or something.

File: src/osd/OSD.cc [hunk 5]
```diff
@@ -6264,23 +6262,46 @@ int OSD::probe_smart_device(const char *device, int timeout, std::string *result
...
 
   bufferlist output;
   ret = output.read_fd(smartctl.get_stdout(), 100*1024);
-  if (ret < 0) {
-    derr << "failed read from smartctl: " << cpp_strerror(-ret) << dendl;
+  // read failed, or immediately returned EOF; 
+  // or sudo failed, and since it writes to stderr, reading from stdout will return 0.
+  if (ret <= 0) { 
+    derr << "failed to read from smartctl: " << cpp_strerror(-ret) << dendl;
+    *result = "{\"error\": \"failed to read from smartctl\", \"exit_code\": \"" + std::to_string(ret) + "\"}";
```
Comment (hsiang41) [inline src/osd/OSD.cc]: Can catch stderr output to clear the command fault reason?

**Comment by yuriw** [comment]
[wip-yuri-testing-2018-07-20-1607](https://trello.com/c/UYqB4xU2/350-wip-yuri-testing-2018-07-20-1607)

**Comment by yuriw** [comment]
Per @neha-ojha rebiew "http://pulpito.ceph.com/yuriw-2018-07-21_15:40:02-rados-wip-yuri-testing-2018-07-20-1607-distro-basic-smithi/2806429/ seems related to osd: add smartctl error output to device map"

**Comment by liewegas** [comment]
maybe, i'm also seeing a mgr crash on master from show-health-metrics, trying to figure out when it started happening.

**Comment by hsiang41** [comment]
In my testing, I add code to output stderr message as below:
error output is: "sudo: sorry, you must have a tty to run sudo".
I modify the /etc/sudoers, and marke below string to solve this problem.
#Defaults    requiretty
If this code add stderr output that can easy to know what's happen.
Like as:
    bufferlist err;
    err.read_fd(smartctl.get_stderr(), 100 * 1024);

This issue maybe same as my.

**Comment by neha-ojha** [comment]
needs rebase

**Comment by liewegas** [comment]
This is resolved in master by 2fb6315db135830a08ac2e3987c7dcfdeb09ca1b and related commits.

**Pull request closed without merging**